Demo/setup (#4)
kokkos/kokkos-core-wiki: README.md

########################################################
#			Welcome to the Kokkos Documentation Workflow     #
########################################################

	Here, we summarize the steps for setting up the kokkos-core-wiki repo, and a
branch-based workflow for updating documentation in the main Kokkos project.  The kokkos-core-wiki is the
canonical source of Kokkos documentation "Truth".

	CAVEAT: Creation of wholly new pages will require Review and Approval in some

> cases.  Check with your team lead, if you are uncertain.

Reviewer: This repository is also meant for external (non-Kokkos, non-Sandia) contributors for whom this paragraph sounds confusing/not encouraging to participate.

	STEP 1) Create your own fork of the kokkos-core-wiki repo by navigating to
					https://github.com/kokkos/kokkos-core-wiki, and clicking "Fork" in the upper
					right of the github page.

	STEP 2) Clone your forked repo:

					git clone git@github.com:ajpowelsnl/kokkos-core-wiki.git

	STEP 3) Create a topic branch in your local fork of the kokkos-core-wiki:

> 					git checkout -b demo/setup

Reviewer: I would differentiate between one-time setup steps and the ones that you need to do for every pull request. This one should be before the current "STEP 10)" IMHO.

	STEP 4) Check your remotes:

					[ajpowel@kokkos-dev-2 kokkos-core-wiki]$ git remote -v
					origin	git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
					origin	git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)

	STEP 5) Rename your (local) repo to disambiguate:

					git remote rename origin ajpowelsnl

	STEP 6) Add the remote kokkos-core-wiki repo (i.e., the repo you will create a PR on):

> 					git remote add origin https://github.com/kokkos/kokkos-core-wiki.git

Author: OK, I can update the instructions to do it correctly the first time.

	STEP 7) Add the Kokkos project Wiki from your fork (you will push changes to this repo to preview): 

>  					git remote add my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git

Reviewer: For this to work, you have to set up the Wiki in the fork first, e.g., by just creating a dummy page.

	STEP 8) Verify your remotes (you should now have three different repos: 
					your local, forked kokkos-core-wiki ("ajpowelsnl" below), the remote kokkos-core-wiki ("origin" below),
					and a local main-project Wiki associated with your forked Kokkos project ("my_wiki" below)):

				git remote -v
	
				ajpowelsnl	git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
				ajpowelsnl	git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
				my_wiki	git@github.com:ajpowelsnl/kokkos.wiki.git (fetch)
				my_wiki	git@github.com:ajpowelsnl/kokkos.wiki.git (push)
				origin	https://github.com/kokkos/kokkos-core-wiki.git (fetch)
				origin	https://github.com/kokkos/kokkos-core-wiki.git (push)
 
	STEP 9) Set up your local topic branch track origin/main:

 				git checkout main

> 				git branch --set-upstream-to=origin/main 

Reviewer: Is there a good reason to even care about the local `main` branch? I would just rebase on top of `origin/main`.

> 				git pull origin/main

Reviewer: ```suggestion
				git pull origin
```
I am not aware that you can pull an individual branch like this. I get
```
$ git pull upstream/main
fatal: 'upstream/main' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

				git show
				commit f0e10f59f50f23930a2f4eedfce3eed869344277 (HEAD -> main, origin/main)
				Merge: 7b5063e ec9f444
				Author: Damien L-G <dalg24@gmail.com>
				Date:   Mon Feb 28 21:09:11 2022 -0500

						Merge pull request #3 from dalg24/fixup_gh_actions
						
						Fixup path gh workflow

	STEP 10) Update your topic branch,rebasing on origin/main:
				
				git checkout demo/setup

> 				git rebase main

Reviewer: ```suggestion
				git rebase origin/main
```

	STEP 10) Make the desired changes (on your local topic branch), and push

> 				those changes to the main project Wiki (of your fork).  Nota bene: you

Reviewer: ```suggestion
				those changes to the your `kokkos` fork's Wiki.  Nota bene: you
```

				will need to use the "-f" option for the **FIRST** push; this option will overwrite the file.  

				git push -f my_wiki demo/setup:master

	STEP 11) Preview your changes by navigating to the main project Wiki of your
				fork:

				https://github.com/ajpowelsnl/kokkos/wiki

	STEP 12) If your previewed changes are good, push your local topic branch to
				your fork of the kokkos-core-wiki to create a pull request.  Please
				note that pushes to the main branch will be automatically, immediately

> 				deployed to the Kokkos Wiki:

Reviewer: This shouldn't be possible in the first place, right? `main` is protected.

# Kokkos Documentation Workflow <br/><br/>

Here, we summarize the steps for setting up the kokkos-core-wiki repo, and a
branch-based workflow for updating documentation in the main Kokkos project.  The kokkos-core-wiki is the
canonical source of Kokkos documentation "Truth".<br/>

Sandia employees: Creation of wholly new pages will require Review and Approval (R&A).  Check with your Sandia team lead, if you have questions about the R&A process.<br/><br/>

* **Step 1) Create your own fork of the kokkos-core-wiki repo by navigating to `https://github.com/kokkos/kokkos-core-wiki`, and clicking "Fork" in the upper right of the github page**.<br/>

* **Step 2) Clone your forked repo, and add it as a remote:**<br/>

```

> git clone git@github.com:ajpowelsnl/kokkos-core-wiki.git

Reviewer: This already creates a remote called `origin`. You should `clone` the repository you want to be called `origin`.
In other words, if you want `origin` to point to `kokkos/kokkos-core-wiki` and `ajpowelsnl` to `ajpowelsnl/kokkos-core-wiki` do
```
git clone git@github.com:kokkos/kokkos-core-wiki.git
git remote add ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git
```
In other words, if you want `upstream` to point to `kokkos/kokkos-core-wiki` and `origin` to `ajpowelsnl/kokkos-core-wiki` do
```
git clone git@github.com:ajpowelsnl/kokkos-core-wiki.git
git remote add upstream git@github.com:kokkos/kokkos-core-wiki.git
```
Of course, there is also an `--origin <name>` option you specify while cloning to give the remote created a different name than `origin` (untested):
```
git clone --origin ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git
git remote add origin git@github.com:kokkos/kokkos-core-wiki.git
```

git remote add ajpowelsnl
git@github.com:ajpowelsnl/kokkos-core-wiki.git
```

* **Step 3) Add the remote repo as "origin" or "upstream", according to your preference.  You will create pull requests against this repo:**<br/>

```
git remote add origin git@github.com:kokkos/kokkos-core-wiki.git
```

* **Step 4) Check your remotes; you should see both your fork (`ajpowelsnl`) and the main project repo (`origin`):**<br/>

```
[ajpowel@kokkos-dev-2 kokkos-core-wiki]$ git remote -v
ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
origin https://github.com/kokkos/kokkos-core-wiki.git (fetch)
origin https://github.com/kokkos/kokkos-core-wiki.git (push)
```

* **Step 5) Create a topic branch in your local fork of the kokkos-core-wiki:**<br/>

```
git checkout -b demo/setup
```

* **Step 6) Add the Kokkos project Wiki repo from your fork as a remote (you will push changes to this repo to preview before creating a pull request on `origin git@github.com:kokkos/kokkos-core-wiki.git`):**<br/>

```
git remote add my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git
```

> 

Reviewer: Before doing this, you should make sure that the `Wiki` is not empty. At least, I couldn't push to or pull from this repository before creating a dummy page.

```
[ajpowel@kokkos-dev-2 kokkos-core-wiki]$ git remote -v
ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git (fetch)
my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git (push)
origin https://github.com/kokkos/kokkos-core-wiki.git (fetch)
origin https://github.com/kokkos/kokkos-core-wiki.git (push)
```

* **Step 8) Set up your local topic branch to track origin/main:**<br/>

```
git checkout main
git branch --set-upstream-to=origin/main 
git pull origin/main
git show
commit f0e10f59f50f23930a2f4eedfce3eed869344277 (HEAD -> main, origin/main)
Merge: 7b5063e ec9f444
Author: Damien L-G <dalg24@gmail.com>
Date:   Mon Feb 28 21:09:11 2022 -0500

Merge pull request #3 from dalg24/fixup_gh_actions

Fixup path gh workflow
```

* **Step 9) Update your topic branch, rebasing on origin/main:**<br/>

```
git checkout demo/setup
git rebase main

> ```

Reviewer: Again, I don't see much of a reason to bother with the local `main` branch and would just do
```
git fetch origin main
git checkout demo/setup
git rebase origin/main
```
Most importantly,
```
git fetch origin/main
```
should be
```
git fetch origin main
```

> * **Step 10) Make the desired changes (on your local topic branch), and push to the main project Wiki (of your fork).  Nota bene: you will need to use the `-f` option to push; this option will overwrite existing files.**<br/>

Reviewer: ```suggestion
* **Step 10) Make the desired changes (on your local topic branch), and push to the main project Wiki (of your fork).  Nota bene: for your first commit, you will need to use the `-f` option to push; this option will overwrite existing files.**<br/>
```

```
git push -f my_wiki demo/setup:master
```

* **Step 11) Preview your changes by navigating to the main project Wiki of your fork:**<br/>

```
https://github.com/ajpowelsnl/kokkos/wiki
```

> * **Step 12) If your previewed changes are good, push your local topic branch to your fork of the kokkos-core-wiki to create a pull request.  Please note that pushes to your branch will be automatically, immediately deployed to your *fork* of the Kokkos project Wiki.  You cannot push directly to the `main` branch of the remote repo:**<br/>

Reviewer: I find the second sentence confusing. What are you trying to say? We are not pushing to the Wiki of the fork of `kokkos-core-wiki` here but directly to the fork.

```suggestion
* **Step 12) If your previewed changes are good, push your local topic branch to your fork of the kokkos-core-wiki to create a pull request. You cannot push directly to the `main` branch of the remote repo:**<br/>
```

# Kokkos Documentation Workflow <br/><br/>

Here, we summarize the steps for setting up the kokkos-core-wiki repo, and a
branch-based workflow for updating documentation in the main Kokkos project.  The kokkos-core-wiki is the
canonical source of Kokkos documentation "Truth".  It is no longer possible to
directly edit the main Kokkos project Wiki.<br/>

Sandia employees: Creation of wholly new pages will require Review and Approval (R&A).  Check with your Sandia team lead, if you have questions about the R&A process.<br/><br/>

> * Step 1) Create your own fork of the kokkos-core-wiki repo by navigating to `https://github.com/kokkos/kokkos-core-wiki`, and clicking "Fork" in the upper right of the github page.<br/>

Reviewer 2: Note: numbered lists are supported and GitHub can do the counting for you.

```

1. lore
multiline
text

1. ipsum
1. text
```
will give you:

> 
> 1. lore
> multiline
> text
> 
> 1. ipsum
> 1. text

> * **Step 1) Create your own fork of the kokkos-core-wiki repo by navigating to `https://github.com/kokkos/kokkos-core-wiki`, and clicking "Fork" in the upper right of the github page**.<br/>

Reviewer: I wouldn't have the steps in boldface when the code blocks are not.

Author: ok, cool.  I can fix this.

> * Step 2) Clone the remote kokkos-core-wiki repo, and add your fork as a remote:<br/>

Reviewer 2: Upstream/origin convention:
```suggestion
* Step 2) Clone your fork and add upstream `kokkos-core-wiki` repo as a remote:<br/>
```

```

> git clone git@github.com:kokkos/kokkos-core-wiki.git

Reviewer 3: Use upstream/origin convention:
```suggestion
git clone git@github.com:ajpowelsnl/kokkos-core-wiki.git
```

> git remote add ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git

Reviewer 3: ```suggestion
git remote add upstream git@github.com:kokkos/kokkos-core-wiki.git
```

```

> * Step 3) Check your remotes; you should see both your fork (`ajpowelsnl`) and the main kokkos-core-wiki project repo (`origin`):<br/>

Reviewer 3: ```suggestion
* Step 3) Check your remotes; you should see both your fork (`origin`) and the main kokkos-core-wiki project repo (`upstream`):<br/>
```

```
[ajpowel@kokkos-dev-2 kokkos-core-wiki]$ git remote -v

> ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
> ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
> origin https://github.com/kokkos/kokkos-core-wiki.git (fetch)
> origin https://github.com/kokkos/kokkos-core-wiki.git (push)

Reviewer 3: ```suggestion
origin git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
origin git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
upstream https://github.com/kokkos/kokkos-core-wiki.git (fetch)
upstream https://github.com/kokkos/kokkos-core-wiki.git (push)
```

```

* Step 4) Create a topic branch in your local fork of the kokkos-core-wiki:<br/>

```
git checkout -b demo/setup
```

> * Step 5) Add your fork's  *Kokkos project Wiki* repo as a remote (you will push changes to this repo to preview before creating a pull request on `origin git@github.com:kokkos/kokkos-core-wiki.git`).  If your *Kokkos project Wiki* is empty, then you will need to create a first page.  To create this first page, navigate to the Wiki in the banner (*e.g.*, `https://github.com/ajpowelsnl/kokkos1-kernels/wiki`), and click the green button that says, `Create the first page`.  Once you have created this first page, you should now be able to add the Wiki repo as a remote:<br/>

Reviewer 3: ```suggestion
* Step 5) Add your fork's  *Kokkos project Wiki* repo as a remote (you will push changes to this repo to preview before creating a pull request on `upstream git@github.com:kokkos/kokkos-core-wiki.git`).  If your *Kokkos project Wiki* is empty, then you will need to create a first page.  To create this first page, navigate to the Wiki in the banner (*e.g.*, `https://github.com/ajpowelsnl/kokkos1-kernels/wiki`), and click the green button that says, `Create the first page`.  Once you have created this first page, you should now be able to add the Wiki repo as a remote:<br/>
```

```
git remote add my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git
```

> * Step 6) Verify your remotes.  You should now have three different repos: your local, forked kokkos-core-wiki (`ajpowelsnl`), the remote kokkos-core-wiki (`origin`), and the Wiki associated with your forked Kokkos project (`my_wiki`):<br/>

Reviewer 3: ```suggestion
* Step 6) Verify your remotes.  You should now have three different repos: your local, forked kokkos-core-wiki (`origin`), the remote kokkos-core-wiki (`upstream`), and the Wiki associated with your forked Kokkos project (`my_wiki`):<br/>
```

```

> [ajpowel@kokkos-dev-2 kokkos-core-wiki]$ git remote -v
> ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
> ajpowelsnl git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
> my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git (fetch)
> my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git (push)
> origin https://github.com/kokkos/kokkos-core-wiki.git (fetch)
> origin https://github.com/kokkos/kokkos-core-wiki.git (push)

Reviewer 3: ```suggestion
[ajpowel@kokkos-dev-2 kokkos-core-wiki]$ git remote -v
origin git@github.com:ajpowelsnl/kokkos-core-wiki.git (fetch)
origin git@github.com:ajpowelsnl/kokkos-core-wiki.git (push)
my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git (fetch)
my_wiki git@github.com:ajpowelsnl/kokkos.wiki.git (push)
upstream https://github.com/kokkos/kokkos-core-wiki.git (fetch)
upstream https://github.com/kokkos/kokkos-core-wiki.git (push)
```

```

> * Step 7) Fetch new changes to the remote (`origin`), and rebase on your topic branch (`demo/setup`) to update:<br/>

Reviewer 3: ```suggestion
* Step 7) Fetch new changes to the remote (`upstream`), and rebase on your topic branch (`demo/setup`) to update:<br/>
```

```

> git fetch origin main
> git checkout demo/setup
> git rebase origin/main

Reviewer 3: ```suggestion
git fetch upstream main
git checkout demo/setup
git rebase upstream/main
```

```

* Step 8) Make the desired changes (on your local topic branch), and push to the main project Wiki (of your fork).  *Nota bene*: for your first commit, you will need to use the `-f` option to push; this option will overwrite existing files.<br/>

```
git push -f my_wiki demo/setup:master
```

* Step 9) Preview your changes by navigating to the main project Wiki of your fork:<br/>

```
https://github.com/ajpowelsnl/kokkos/wiki
```

* Step 10) If your previewed changes are good, push your local topic branch to your fork of the kokkos-core-wiki repo to create a pull request.  You cannot push directly to the `main` branch of the remote repo:<br/>

```

> git push -f ajpowelsnl demo/setup 

Reviewer 3: ```suggestion
git push -f origin demo/setup 
```

> git push -f ajpowelsnl demo/setup 

Reviewer 2: Simple `git push` should be enough here:
```suggestion
git push
```

Reviewer 2: Note: on the first try the message will be:
```
fatal: The current branch demo/setup has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin demo/setup
```
but I don't think we should go that deep into Git basics - `--set-upstream` is suggested in the error message.
